cranelift/codegen/src/machinst/mod.rs

```diff
@@ -164,14 +167,31 @@ pub trait MachInst: Clone + Debug {
     /// Generate a move.
     fn gen_move(to_reg: Writable<Reg>, from_reg: Reg, ty: Type) -> Self;
 
-    /// Generate a constant into a reg.
+    /// Determine whether a constant value should be generated in place. Higher-level users can use
+    /// this to determine whether `gen_constant` or `gen_constant_in_pool` should be used.
```

Comment:
s/in place/into a constant pool/, no? (Or perhaps "should be generated into a constant pool and loaded (if true); otherwise, it should be generated in place by an instruction or instruction sequence."

We should also clarify that if this function returns `false`, then `gen_constant` must be able to handle this value, and if `true`, then `gen_constant_in_pool` should.

Reply:
Looks like there's a solution involving more traits that could avoid the assertions and the dummy unimplemented:

- have a trait `ConstantGenerator` that has a method "gen_constant"
- have a trait `PooledConstantGenerator`, that has three methods "gen_constant_without_pool", "gen_constant_with_pool", "must_generate_in_pool"
- then `PooledConstantGenerator` gets an blanket impl of `ConstantGenerator` based on these three functions above.

Then backends have to implement `ConstantGenerator` either directly, or via an opt-in choice to use the pooled variant. It's a bit more involved though...

Reply:
I like this scheme -- @abrown, does this seem reasonable to incorporate in your PR? I'd be happy to review it as a refactoring followup otherwise.

Reply:
I like it in principle! Let me dive back in and see how it comes out in practice.